app/code/core/Mage/Csp/Model/Observer/Abstract.php

```diff
@@ -0,0 +1,84 @@
+<?php
+
+declare(strict_types=1);
+
+/**
+ * @copyright  For copyright and license information, read the COPYING.txt file.
+ * @link       /COPYING.txt
+ * @license    Open Software License (OSL 3.0)
+ * @package    Mage_Csp
+ */
+
+abstract class Mage_Csp_Model_Observer_Abstract
+{
+    /**
+     * Common method to add CSP headers for a specific area
+     *
+     * @param Mage_Core_Model_App_Area::AREA_FRONTEND|Mage_Core_Model_App_Area::AREA_ADMINHTML $area
+     * @throws Zend_Controller_Response_Exception
+     */
+    protected function addCspHeaders(Varien_Event_Observer $observer, string $area): void
+    {
+        /** @var Mage_Core_Controller_Response_Http $response */
+        $response = $observer->getEvent()->getDataByKey('response');
```

Comment:
@sreichel Why in your patch do you want to use getDataByKey('response') instead of getResponse()?

Reply:
`getResponse()` is a magic method that finally calls Varien_Object::getData() with some unnecessary checks. getDataByKey() is avoids some overhead.

Reply:
It's an unnecessary over-optimization. I prefer to keep clearer and more fluent code. I haven't found other code like this. In observers, like in `Mage_Wishlist_Model_Observer::processAddToCart()`, `getResponse()` is always used. Maybe in the future, a magic method `getResponse()` could return the correct type as well.

Reply:
getDataByKey() comes from M2 and was added 8 month ago, so it's only used in a few places atm. But I've added it to avoid that overhead. It's not an over-microoptimization, it's clearly faster.